File: haskell/app/Day06.hs
module Day06 (main, slice, allUnique, findMarker) where

import qualified Data.Set as Set
import Helpers (yoloReadFile)

inputData :: String
inputData = yoloReadFile "../data/input06.txt"

slice :: Int -> Int -> String -> String
slice from to xs = take (to - from + 1) (drop from xs)

allUnique :: String -> Bool
allUnique str = Set.size (Set.fromList str) == length str

findMarker :: String -> Int -> Maybe (String, Int)
findMarker str seqLen = findMarker' str seqLen seqLen

findMarker' :: String -> Int -> Int -> Maybe (String, Int)
findMarker' str seqLen index
  | index >= length str = Nothing
  | allUnique chunk = Just (chunk, index)
  | otherwise = findMarker' str seqLen (index + 1)
  where chunk = slice (index - seqLen) (index - 1) str

main :: IO ()
main = print (packetMarker, messageMarker)
  where packetMarker = findMarker inputData 4
        messageMarker = findMarker inputData 14

File: typescript/src/day06.ts
import { readFileSync } from 'fs';

const inputFilename: string = '../data/input06.txt';
const inputData = readFileSync(inputFilename, 'utf8');
// const sampleData = 'mjqjpqmgbljsphdztnvjfqwrcgsmlb';

const allUnique = (str: string) => {
  let chars: any = {};
  for (let i = 0; i < str.length; i++) {
    const char = str[i];
    if (chars[char]) {
      return false;
    }
    chars[char] = true;
  }
  return true;
};

const findMarker = (data: string, seqLength: number) => {
  for (let i = seqLength; i < data.length; i++) {
    let chars = data.slice(i - seqLength, i);
    if (allUnique(chars)) {
      return i;
    }
  }
  return null;
};

console.log({
  'packetMarker at': findMarker(inputData, 4),
  'messageMarker at': findMarker(inputData, 14),
});
